python/ray/autoscaler/_private/autoscaler.py

```diff
@@ -478,50 +504,65 @@ def _get_nodes_allowed_to_terminate(
             config's available_node_types dict.
         (b) Deleting the node would violate the min_workers constraint for that
             worker's node_type.
 
-        Also updates and returns the dictionary of node type counts.
+        Returns KeepOrTerminate.terminate when either the node type is not
+        among available node type, or both the following hold:
+        (a) The worker's node_type is present among the keys of the current
+            config's available_node_types dict.
+        (b) Keeping the node would violate the max_workers constraint for that
+            worker's node_type.
+
+        Return KeepOrTerminate.decide_later otherwise.
+
 
         Args:
             node_type_counts(Dict[NodeType, int]): The non_terminated node
                 types counted so far.
         Returns:
-            bool: True if the node should be kept. False otherwise.
-            Dict[NodeType, int]: Updated node type counts
+            KeepOrTerminate: keep if the node should be kept, terminate if the
+            node should be terminated, decide_later if we are allowed
+            to terminate it, but do not have to.
+            Optional[str]: reason for termination.
         """
-        new_node_type_counts = copy.deepcopy(node_type_counts)
         tags = self.provider.node_tags(node_id)
         if TAG_RAY_USER_NODE_TYPE in tags:
             node_type = tags[TAG_RAY_USER_NODE_TYPE]
-            if node_type not in self.available_node_types:
-                # The node type has been deleted from the cluster config.
-                # Don't keep the node.
-                return False, new_node_type_counts
-            new_node_type_counts[node_type] += 1
-            min_workers = self.available_node_types[node_type].get(
+            head_node_type = self.config["head_node_type"]
+
+            min_workers = self.available_node_types.get(node_type, {}).get(
                 "min_workers", 0)
-            max_workers = self.available_node_types[node_type].get(
+            max_workers = self.available_node_types.get(node_type, {}).get(
                 "max_workers", 0)
-            if new_node_type_counts[node_type] <= min(min_workers,
-                                                      max_workers):
-                return True, new_node_type_counts
-
-        return False, new_node_type_counts
+            if node_type not in self.available_node_types:
+                # The node type has been deleted from the cluster config.
+                # Allow terminating it if needed.
+                available_node_types = list(self.available_node_types.keys())
+                return (KeepOrTerminate.terminate,
+                        f"not in available_node_types: {available_node_types}")
+            new_count = node_type_counts[node_type] + 1
+            if (new_count <= min(min_workers, max_workers)
+                    or node_type == head_node_type):
```

Comment:
I don't think the extra logic for the head node-type is needed here (though might be misunderstanding).

Note that it isn't possible for node_id to represent the head node itself, as node_id is an element of self.workers(), which is the list of worker (non-head) nodes. node_id could be a worker of head_node_type, but it's not the head.

"worker of head node-type" is such a slippery notion that we usually recommend setting min_workers=max_workers=0 for the head node-type.

Reply:
without this special handling, i see some nodes of `empty_node` type removed (max_workers_per_type) in (new) testScaleDownMaxWorkers test.
is there  problem with how the test is setup?

Reply:
Probably missed a detail -- but why was the test spawning workers of type empty_node? It looks like max_workers is set to 0 for that type.

Reply:
the issue was with the way I set the test up (there was no head node and this autoscaler was spawning one, and the killing it, and the re-spawnign it.